source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -682,7 +682,14 @@ Network::FilterStatus Filter::onNewConnection() {
   return establishUpstreamConnection();
 }
 
-bool Filter::startUpstreamSecureTransport() { return upstream_->startUpstreamSecureTransport(); }
+bool Filter::startUpstreamSecureTransport() {
+  bool switched_to_tls = upstream_->startUpstreamSecureTransport();
```

Comment:
If you modify signature of this method as described above, you can get access to Pool's connection, extract ssl and  set it in upstream_info.